cpp/comms/std/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,408 @@
+    } else {
+      ucp_ep_v[i] = nullptr;
+    }
+  }
+
+  inject_comms(*handle, comm, (ucp_worker_h)ucp_worker, eps_sp, size, rank);
+
+#else
+  inject_comms(*handle, comm, size, rank);
+#endif
+}
+
+void ncclUniqueIdFromChar(ncclUniqueId *id, char *uniqueId, int size) {
+  memcpy(id->internal, uniqueId, size);
+}
+
+/**
+ * @brief Returns a NCCL unique ID as a character array. PyTorch
+ * uses this same approach, so that it can be more easily
+ * converted to a native Python string by Cython and further
+ * serialized to be sent across process & node boundaries.
+ *
+ * @returns the generated NCCL unique ID for establishing a
+ * new clique.
+ */
+void get_unique_id(char *uid, int size) {
+  ncclUniqueId id;
+  ncclGetUniqueId(&id);
+
+  memcpy(uid, id.internal, size);
+}
+
+#ifdef WITH_UCX
+cumlStdCommunicator_impl::cumlStdCommunicator_impl(
+  ncclComm_t comm, ucp_worker_h ucp_worker, std::shared_ptr<ucp_ep_h *> eps,
+  int size, int rank)
+  : _nccl_comm(comm),
+    _ucp_worker(ucp_worker),
+    _ucp_eps(eps),
+    _size(size),
+    _rank(rank),
+    _next_request_id(0) {
+  initialize();
+}
+#endif
+
+cumlStdCommunicator_impl::cumlStdCommunicator_impl(ncclComm_t comm, int size,
+                                                   int rank)
+  : _nccl_comm(comm), _size(size), _rank(rank) {
+  initialize();
+}
+
+void cumlStdCommunicator_impl::initialize() {
+  cudaStreamCreate(&stream);
+
+  CUDA_CHECK(cudaMalloc((void **)&sendbuff, sizeof(int)));
+  CUDA_CHECK(cudaMalloc((void **)&recvbuff, sizeof(int)));
+}
+
+
+
+cumlStdCommunicator_impl::~cumlStdCommunicator_impl() {
+
+  cudaStreamDestroy(stream);
+
+  CUDA_CHECK_NO_THROW(cudaFree(sendbuff));
+  CUDA_CHECK_NO_THROW(cudaFree(recvbuff));
+}
+
+int cumlStdCommunicator_impl::getSize() const { return _size; }
+
+int cumlStdCommunicator_impl::getRank() const { return _rank; }
+
+std::unique_ptr<MLCommon::cumlCommunicator_iface>
+cumlStdCommunicator_impl::commSplit(int color, int key) const {
+  // Not supported by NCCL
+  ASSERT(
+    false,
+    "ERROR: commSplit called but not supported in this comms implementation.");
+}
+
+void cumlStdCommunicator_impl::barrier() const {
+  CUDA_CHECK(cudaMemsetAsync(sendbuff, 0, sizeof(int), stream));
+  CUDA_CHECK(cudaMemsetAsync(recvbuff, 0, sizeof(int), stream));
+
+  allreduce(sendbuff, recvbuff, 1, MLCommon::cumlCommunicator::INT,
+            MLCommon::cumlCommunicator::SUM, stream);
+
+  cudaStreamSynchronize(stream);
+}
+
+void cumlStdCommunicator_impl::isend(const void *buf, int size, int dest,
+                                     int tag, request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  request_t req_id;
+  if (_free_requests.empty())
+    req_id = _next_request_id++;
+  else {
+    auto it = _free_requests.begin();
+    req_id = *it;
+    _free_requests.erase(it);
+  }
+
+  ucp_tag_t ucp_tag = (ucp_tag_t)tag;
```

Comment:
Is `ucp_tag` used somewhere?

Reply:
Nope. The logic in `ucp_help.h` was originally in this class. This got left around in the refactor.